### PR TITLE
minor: Improve summary javadoc check method implementation

### PR DESCRIPTION
**Pull request by AkMo3**
As a supplement to #9174, this PR improves the implementation of some methods.

Regression reports are not required as none of the projects use the summary tag.

Diff Regression config: https://raw.githubusercontent.com/AkMo3/akmo3.github.io/master/SummaryJavadoc/SummaryJavadocConfig.xml

Diff Regression projects: https://raw.githubusercontent.com/AkMo3/akmo3.github.io/master/SummaryJavadoc/openjdk10.properties
 
Report: https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/0ef2523_2021045510/reports/diff/index.html (Report is Clean)

### CODE REVIEW

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java
Comment (nmancus1) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java:365]: It seems like there is a bit of duplicated logic here (from the `getInlineTagNodeWithinHtmlElement()` method) that is being obfuscated by the use of `Optional`. Couldn't you just do something like:

```java

        final Predicate<DetailNode> childPredicate = child ->
                child.getType() == JavadocTokenTypes.JAVADOC_INLINE_TAG
                        || child.getType() == JavadocTokenTypes.HTML_ELEMENT;

        final Optional<DetailNode> node = Arrays.stream(javadoc.getChildren())
                .filter(childPredicate)
                .findFirst();

        return node.map(SummaryJavadocCheck::getInlineTagNodeWithinHtmlElement);

```

?

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java
Comment (nmancus1) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java:407]: Using `anyMatch()`, with Java streams gives no guarantee regarding the order of traversal,  this method could now return true if ANY tag is a summary tag, not just the first.

**Review by nmancus1 — CHANGES_REQUESTED** [review]
Few thoughts:

**Comment by AkMo3** [comment]
@nmancus1 I know it is weird but OpenJdk does not uses `@summary` as it should be. It has been implemented in the project in the below format.
```
test/sun/nio/cs/CheckCaseInsensitiveEncAliases.java:   @summary Test to validate case insensitivity of encoding alias names
test/sun/nio/cs/ISCIITest.java: * @summary Performs baseline and regression test on the ISCII91 charset
test/sun/nio/cs/TestUTF_16.java:   @summary Check correctness of the UTF-16 converter in all its flavors
test/sun/nio/cs/CheckHistoricalNames.java: * @summary Checks canonical names match between old and (NIO) core charsets
```

We are not looking for these, `@summary` should be an INLINE_JAVADOC_TAG, and the search result for such tag is
```
akmo@akmo-Modern-14-B4MW  ~/GitHub/contribution/checkstyle-tester/repositories/openjdk9   master ±  grep -R "{@summary" | wc -l
0
```

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java
Comment (AkMo3) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java:407]: I don't think `CUSTOME_NAME` or rather any other secondary tag can be present in children except the first one because our grammar defines only the first and tag rest are considered text or description. What do you think?
```
javadocInlineTag:
      JAVADOC_INLINE_TAG_START
      (
            CODE_LITERAL (WS | NEWLINE | LEADING_ASTERISK | text)*
            | DOC_ROOT_LITERAL (WS | NEWLINE | LEADING_ASTERISK)*
            | INHERIT_DOC_LITERAL (WS | NEWLINE | LEADING_ASTERISK)*
            | LINK_LITERAL (WS | NEWLINE | LEADING_ASTERISK)+ reference (WS | NEWLINE)*
                ((WS | NEWLINE) description)?
            | LINKPLAIN_LITERAL (WS | NEWLINE | LEADING_ASTERISK)+ reference (WS | NEWLINE)*
                ((WS | NEWLINE) description)?
            | LITERAL_LITERAL (WS | NEWLINE | LEADING_ASTERISK | text)*
            | VALUE_LITERAL (WS | NEWLINE | LEADING_ASTERISK)* ((WS | NEWLINE) reference)?
            | CUSTOM_NAME (WS | NEWLINE | LEADING_ASTERISK)* ((WS | NEWLINE) description)?
      )
      JAVADOC_INLINE_TAG_END
      ;
```

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java
Comment (AkMo3) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java]: Looks great. Thanks a lot.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java
Comment (nmancus1) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java:404]: > What do you think?

I think that the javadoc states that this method will `@return {@code true} if first tag is summary tag.` , but that your changes no longer guarantee that :)

Comment (AkMo3) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java:404]: Can you please explain in a bit? In what cases it can fail? Since we know that there will only be a single `JAVADOC_INLINE_TAG` present, how would a case exits where the type is `CUSTOM_NAME`.

Comment (AkMo3) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java:404]: > I just mentioned this because it is in the javadoc description, and the old code made it very apparent that we were getting the first element in the array.

True, it was visually clear in the previous code, but I thought to make it a bit redundant by not hard-coding things out. But the javadoc still ought to remain same as we still are directly checking the first `INLINE_JAVADOC_TAG`.

Comment (AkMo3) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java:404]: > I just mentioned this because it is in the javadoc description, and the old code made it very apparent that we were getting the first element in the array.

True, it was visually clear in the previous code, but I thought to make it a bit redundant by not hard-coding things out. But the javadoc still ought to remain same as we still are directly checking the first INLINE_JAVADOC_TAG. Please resolve things out if you think the problem is solved.

Comment (nmancus1) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java:404]: > In what cases it can fail?

In practice, probably never :)

What I am saying is that this sort of stream usage is probably overkill, and using a non-deterministic API to get something that is supposed to be in an exact index seems like a bad idea to me. If you want to clean up the code a bit, why not just do:

```java

        final DetailNode firstChild = javadoc.getChildren()[1];
        return firstChild.getType() == JavadocTokenTypes.CUSTOM_NAME
                && SUMMARY_TEXT.equals(firstChild.getText());

```
or

```java
        final DetailNode firstChild = Arrays.asList(javadoc.getChildren()).get(1);
        return firstChild.getType() == JavadocTokenTypes.CUSTOM_NAME
                && SUMMARY_TEXT.equals(firstChild.getText());

```

?

This 2-3 lines of code is concise, easy to debug, and makes the intention very clear. I do not think that this stream usage is beneficial here.

**Comment by nmancus1** [comment]
@AkMo3 I forgot that this inline tag appeared in JDK10, not 9, apologies:

```bash

➜  jdk10 git:(master) grep -R "{@summary" | wc -l
21

```

I know that we cannot run normal regression report (with regular checks) on JDK10+, but I wonder if we could run it with a Javadoc check.  We SHOULD be running regression on real code for any changes to checks, IMO.  If we have no real code to parse, why should such checks exist? :) 

This is an oversimplification, but there must be some project that uses this that we can test on. I would not let it block this PR, though.

**Comment by AkMo3** [comment]
GitHub, generate report

**Comment by AkMo3** [comment]
> https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/3377a06_2021103635/reports/diff/index.html

Report not clear. Moving back to the first implementation.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java
Comment (AkMo3) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java]: Let us not change it then, as it is not a big change.

**Comment by AkMo3** [comment]
GitHub, generate report

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java
Comment (AkMo3) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java]: This change was causing an error, so had to revert back to my previous implementation with a slight change.

**Comment by nmancus1** [comment]
Why don't we have 21 violations? https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/c1d0fc3_2021105229/reports/diff/openjdk10/index.html

**Comment by AkMo3** [comment]
> Just curious, why don't we have 21 violations? https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/c1d0fc3_2021105229/reports/diff/openjdk10/index.html

Because as you see can see in https://github.com/checkstyle/checkstyle/pull/10019#issuecomment-844969094 , there are 5 occuring of `{@summary` which are not in `.java` file

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java
Comment (AkMo3) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java]: @nmancus1 Please resolve if no more improvements.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java
Comment (nmancus1) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java]: Resolved.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java
Comment (nmancus1) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java]: Resolved.

**Review by nmancus1 — APPROVED** [review]

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java
Comment (pbludov) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java:361]: It won't work that way.
For this Javadoc
```
<p>@first<p>
@second
```
The code in master will find `@first`, after your PR this will be `@second`.

**Review by pbludov — CHANGES_REQUESTED** [review]

**Comment by AkMo3** [comment]
GitHub, generate report

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java
Comment (AkMo3) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java]: Done. Please check why the wrecker is failing

Comment (nmancus1) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java]: Wercker is failing on `./.ci/wercker.sh no-error-hibernate-search`:

```bash
[WARNING] Rule 1: org.apache.maven.plugins.enforcer.RequireMavenVersion failed with message:
Detected Maven Version: 3.6.3 is not in the allowed range 3.8.1.

```

Comment (AkMo3) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java]: Is this failure related to my change?

Comment (nmancus1) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java]: No, hibernate just bumped their minimum maven version. We will probably need to open an issue in their Jira.

Comment (nmancus1) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java]: @AkMo3 please rebase.

**Comment by AkMo3** [comment]
Github, rebase

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java
Comment (pbludov) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java:357]: Predicates are difficult to reuse. Please extract this predicate to a separated method.
This method, `getFirstInlineTag', on the contrary, can be inlined into `containsSummaryTag`.

**Review by pbludov — CHANGES_REQUESTED** [review]
One minor please.

**Review by pbludov — APPROVED** [review]
I'm ok to merge if CI passed.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java
Comment (AkMo3) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java]: Done

**Comment by AkMo3** [comment]
GitHub, generate report

**Review by romani — APPROVED** [review]

**Pull request merged**
